src/core/diffusion_grid.h

```diff
@@ -559,6 +565,204 @@ class DiffusionGrid {
     c1_.swap(c2_);
   }
 
+  void RK() {
+    // check if diffusion coefficient and decay constant are 0
+    // i.e. if we don't need to calculate diffusion update
+    if (IsFixedSubstance()) {
+      return;
+    }
+
+    r1_.resize(total_num_boxes_);
+    r2_.resize(total_num_boxes_);
+    r3_.resize(total_num_boxes_);
+
+    const auto nx = num_boxes_axis_[0];
+    const auto ny = num_boxes_axis_[1];
+    const auto nz = num_boxes_axis_[2];
+
+    const double ibl2 = 1 / (box_length_ * box_length_);
+    const double d = 1 - dc_[0];
+    double step = diffusion_step_;
+    double h = dt_ / step;
+#define YBF 16
+    for (size_t i = 0; i < step; i += 1) {
+      for (size_t order = 0; order < 2; order++) {
+#pragma omp parallel for collapse(2)
+        for (size_t yy = 0; yy < ny; yy += YBF) {
+          for (size_t z = 0; z < nz; z++) {
+            size_t ymax = yy + YBF;
+            if (ymax >= ny) {
+              ymax = ny;
+            }
+            for (size_t y = yy; y < ymax; y++) {
+              size_t x = 0;
+              int c, n, s, b, t;
+              c = x + y * nx + z * nx * ny;
+#pragma omp simd
+              for (x = 1; x < nx - 1; x++) {
+                ++c;
+                ++n;
+                ++s;
+                ++b;
+                ++t;
+
+                if (y == 0 || y == (ny - 1) || z == 0 || z == (nz - 1)) {
+                  continue;
+                }
+
+                n = c - nx;
+                s = c + nx;
+                b = c - nx * ny;
+                t = c + nx * ny;
+
+                double h2 = h / 2.0;
+
+                if (order == 0) {
+                  k_[0] = (d * (c1_[c - 1] - 2 * c1_[c] + c1_[c + 1]) * ibl2 +
+                           d * (c1_[s] - 2 * c1_[c] + c1_[n]) * ibl2 +
+                           d * (c1_[b] - 2 * c1_[c] + c1_[t]) * ibl2);
+                  r1_[c] = c1_[c] + (k_[0] * h2);
+                } else if (order == 1) {
+                  k_[1] = (d * (c1_[c - 1] - 2 * c1_[c] + c1_[c + 1]) * ibl2 +
+                           d * (c1_[s] - 2 * c1_[c] + c1_[n]) * ibl2 +
+                           d * (c1_[b] - 2 * c1_[c] + c1_[t]) * ibl2);
+
+                  c2_[c] = c1_[c] + (k_[1] * h);
+                }
+              }
+              ++c;
+              ++n;
+              ++s;
+              ++b;
+              ++t;
+            }  // tile ny
+          }    // tile nz
+        }      // block ny
+      }
+      c1_.swap(c2_);
+    }
+  }
+
+  void RKLeaking() {
+    // check if diffusion coefficient and decay constant are 0
+    // i.e. if we don't need to calculate diffusion update
+    if (IsFixedSubstance()) {
+      return;
+    }
+
+    r1_.resize(total_num_boxes_);
+    r2_.resize(total_num_boxes_);
+    r3_.resize(total_num_boxes_);
```

Comment:
You shouldn't resize here. Resizing should only be done in the Update() call. You would otherwise be unnecessarily resizing if the total number of boxes remains the same over iterations.